src/navigation/navigate.ts

```diff
@@ -0,0 +1,45 @@
+import { navigateToUrl } from "single-spa";
+
+export function navigate({ to }: NavigateOptions): void {
+  const target = interpolateUrl(to);
+  const isSpaPath = target.startsWith(window.getOpenmrsSpaBase());
+  if (isSpaPath) {
+    const spaTarget = target.replace(
+      new RegExp("^" + window.getOpenmrsSpaBase()),
+      ""
+    );
+    navigateToUrl(spaTarget);
+  } else {
+    window.location.assign(target);
+  }
+}
+
+// package-local / "protected"
+export function interpolateUrl(template: string): string {
+  return interpolateString(template, {
+    openmrsBase: window.openmrsBase,
+    openmrsSpaBase: window.getOpenmrsSpaBase()
+  }).replace(/^\/\//, "/"); // remove extra initial slash if present
+}
+
+// package-local / "protected"
+export function interpolateString(template: string, params: object): string {
+  const names = Object.keys(params);
+  const vals = Object.values(params);
+  if (template.includes("`")) {
+    throw Error("Template may not include backticks");
+  }
+  return new Function(...names, `return \`${template}\`;`)(...vals);
```

Comment:
It's nifty, but do we know the perf. result of calling it like that? Maybe we could cache this based on the template string + arg names (then we can just re-use it with different values).

Reply:
I'll measure it, then we can decide if it warrants the complexity of memoizing. What's a good benchmark time to shoot for?

Reply:
Hm I would compare three things

1. The one above
2. Something like `return names.reduce((prev, curr) => prev.split('${' + curr + '}').join(params[curr]), template);`, or maybe a better variation of this
3. Adding caching (to 1)

Reply:
I've added a small jsben.ch run:

https://jsben.ch/iwN6A

Reply:
Quite interesting:

**Chrome**
![image](https://user-images.githubusercontent.com/1766191/82864597-26b19500-9f25-11ea-9d9d-165a4c508a24.png)

**Firefox**
![image](https://user-images.githubusercontent.com/1766191/82864573-1bf70000-9f25-11ea-9be0-3597d377ac41.png)

On Chrome the difference is not so much (but noticable), however, on Firefox things get crazy ...

Reply:
Maybe a combination of (2) and (3) would be super performing?

Reply:
Oh wow, the performance difference for Firefox is huge! Thanks for checking that. I think Code Block 2 is the right approach. Code block 3 seems like overkill for this (I don't expect this function to run more than ~10 times, once, at component mount time).

Reply:
Agreed + it would need to be more complicated than my implementation (a cache without an expiration policy is a memory leak...).

So let's go for option 2. Thanks!